### PR TITLE
feat: extend new pythonrt with remote

### PR DESCRIPTION
**Pull request by efiShtain**
This is a refactor on the new remoteRT
it doesn't change the logic, only split the files in a different way 
so it can be extended to be used with remote runners in a simpler way

### CODE REVIEW

File: backend/runtimes/runtimes.go
Comment (tebeka) [inline backend/runtimes/runtimes.go]:
```suggestion
	EnableRemoteRunner          bool     `koanf:"enable_remote_runner"`
```

File: backend/runtimes/runtimes.go
Comment (tebeka) [inline backend/runtimes/runtimes.go]: Try to fix this with `TestMain` and `AK_WORKER_PYTHON`

File: backend/runtimes/runtimes.go
Comment (tebeka) [inline backend/runtimes/runtimes.go:49]: You can probably do it with fx and a dummy service, IMO not worth it.

File: backend/runtimes/runtimes.go
Comment (tebeka) [inline backend/runtimes/runtimes.go:52]: Validate the `cfg.RemoteRunnerEnpoints` is valid (every one in `host:port` format, maybe even try to reach them).

File: backend/runtimes/runtimes.go
Comment (tebeka) [inline backend/runtimes/runtimes.go]: Can we return error instead of panicking?

File: backend/runtimes/runtimes.go
Comment (tebeka) [inline backend/runtimes/runtimes.go:57]: I prefer:

```go
err = pythonruntime.ConfigureRemoteRunnerManager( .... )
if err != nil {
...
```

File: cmd/ak/cmd/runtimes/runtime.go
Comment (tebeka) [inline cmd/ak/cmd/runtimes/runtime.go:42]: Good questions :)
@itayd?

File: internal/backend/httpsvc/intercept.go
Comment (tebeka) [inline internal/backend/httpsvc/intercept.go]: Maybe a custom mux instead? No sure.

File: internal/backend/temporalclient/config.go
Comment (tebeka) [inline internal/backend/temporalclient/config.go]: Revert this?

File: runtimes/pythonrt/local_runner.go
Comment (tebeka) [inline runtimes/pythonrt/local_runner.go]: I know it's my code, but change to `uuid.NewString()`

File: runtimes/pythonrt/pythonrt_test.go
Comment (tebeka) [inline runtimes/pythonrt/pythonrt_test.go]: No panic, pass `t *testing.T` to this setup and call `t.Fatal`

File: runtimes/pythonrt/pythonrt_test.go
Comment (tebeka) [inline runtimes/pythonrt/pythonrt_test.go]: No panic please.

File: runtimes/pythonrt/pythonrt_test.go
Comment (tebeka) [inline runtimes/pythonrt/pythonrt_test.go]: Log the error (`t.Log`)

File: runtimes/pythonrt/runner/remote_pb2_grpc.py
Comment (tebeka) [inline runtimes/pythonrt/runner/remote_pb2_grpc.py]: Why did we go *down* in version?

File: runtimes/pythonrt/runner_manager_local.go
Comment (tebeka) [inline runtimes/pythonrt/runner_manager_local.go]: Make sure before merging there's no commented code left.

File: runtimes/pythonrt/runner_manager_local.go
Comment (tebeka) [inline runtimes/pythonrt/runner_manager_local.go]: Check that PID is up (you can do `kill 0` to the process to see if it's alive)

File: runtimes/pythonrt/worker_grpc_handler.go
Comment (tebeka) [inline runtimes/pythonrt/worker_grpc_handler.go:95]: Create a `Runner` struct that will have `Get`, hide the lock there.

File: runtimes/pythonrt/worker_grpc_handler.go
Comment (tebeka) [inline runtimes/pythonrt/worker_grpc_handler.go:127]: No error. Also set return gRPC errors. For example:

```
return nil, status.Errorf(codes.Internal, "new function value: %s", err)
```

See `svc.go` in my branch.

**Review by tebeka — APPROVED** [review]
Approving in order not to block you. See my comments.
Also, made some backport fixes from main to my branch, see that you get the.

File: .gitignore
Comment (daabr) [inline .gitignore:113]: I intentionally disabled `MANIFEST` to avoid git warnings about non-Python files. Let me check how to improve this line to apply only to Python.

File: .github/workflows/ci-proto.yml
Comment (daabr) [inline .github/workflows/ci-proto.yml:25]: Is it possible to use Buf instead of these new dependencies?

File: runtimes/pythonrt/py-sdk/autokitteh/google.py
Comment (daabr) [inline runtimes/pythonrt/py-sdk/autokitteh/google.py]: Undo this change - the previous version is the right one

```suggestion
        return match.group(3)
```

File: runtimes/pythonrt/py-sdk/autokitteh/google.py
Comment (daabr) [inline runtimes/pythonrt/py-sdk/autokitteh/google.py]: Undo these changes - the previous version is the right one

```suggestion
    API documentation:
    https://docs.autokitteh.com/integrations/google/calendar/python
```

File: runtimes/pythonrt/testdata/pycache/__pycache__/handler.cpython-312.pyc
Comment (daabr) [inline runtimes/pythonrt/testdata/pycache/__pycache__/handler.cpython-312.pyc:1]: Remove this file from this PR, and make sure .gitignore prevents adding it in the future

File: runtimes/pythonrt/testdata/trigger_webhook.go
Comment (daabr) [inline runtimes/pythonrt/testdata/trigger_webhook.go:1]: I think this shouldn't be a Go program, but I could be wrong, and anyway I won't block the PR on this. However, please add a package docstring (https://go.dev/doc/comment#package) that explains what this program does, and what uses it (make).

File: .github/workflows/ci-proto.yml
Comment (itayd) [inline .github/workflows/ci-proto.yml]: arduino?! looks very suspicious.

File: .github/workflows/ci-proto.yml
Comment (itayd) [reply .github/workflows/ci-proto.yml:25]: seconded, though can be done in a later pr.

File: backend/runtimes/runtimes.go
Comment (itayd) [inline backend/runtimes/runtimes.go:22]: ticket

File: backend/runtimes/runtimes.go
Comment (itayd) [inline backend/runtimes/runtimes.go:19]: ticket?

File: backend/runtimes/runtimes.go
Comment (itayd) [inline backend/runtimes/runtimes.go]: this is a well known port? should it be a constant?

File: backend/runtimes/runtimes.go
Comment (itayd) [inline backend/runtimes/runtimes.go]: ?

File: backend/runtimes/runtimes.go
Comment (itayd) [inline backend/runtimes/runtimes.go:48]: so rethink it. i think it's ok.

File: backend/runtimes/runtimes.go
Comment (itayd) [reply backend/runtimes/runtimes.go]: we should

File: backend/runtimes/runtimes.go
Comment (itayd) [inline backend/runtimes/runtimes.go]: we should just return an error if it fails

File: cmd/ak/cmd/runtimes/runtime.go
Comment (itayd) [reply cmd/ak/cmd/runtimes/runtime.go:42]: not sure i understand the question here. `zap.New(nil)` looks very odd. i think we should do https://github.com/autokitteh/autokitteh/pull/663, but i can redo after merge.

File: internal/backend/httpsvc/intercept.go
Comment (itayd) [inline internal/backend/httpsvc/intercept.go]: @tebeka is this the way?

File: internal/backend/httpsvc/intercept.go
Comment (itayd) [reply internal/backend/httpsvc/intercept.go]: yup.

File: proto/autokitteh/remote/v1/remote.proto
Comment (itayd) [inline proto/autokitteh/remote/v1/remote.proto:11]: ?

File: proto/autokitteh/remote/v1/remote.proto
Comment (itayd) [inline proto/autokitteh/remote/v1/remote.proto:24]: ticket

File: proto/autokitteh/remote/v1/remote.proto
Comment (itayd) [inline proto/autokitteh/remote/v1/remote.proto:1]: ticket. also this file needs some big big big comment explaining what is going on here. what talks to what? when? why?

some kind of ascii diagram maybe.

File: runtimes/pythonrt/build.go
Comment (itayd) [inline runtimes/pythonrt/build.go:17]: this is @tebeka 's code?

File: runtimes/pythonrt/local_runner_test.go
Comment (itayd) [inline runtimes/pythonrt/local_runner_test.go:186]: @tebeka ?

**Comment by itayd** [comment]
initial comments. not sure what's new code here and what just moved. we really need a proper README for this entire thing to explain the design and some basic organization.

I defer to @tebeka as long as this code is enabled by a feature flag, and not just replace what we have now.

File: runtimes/pythonrt/testdata/trigger_webhook.go
Comment (ig-ra) [reply runtimes/pythonrt/testdata/trigger_webhook.go:1]: previously there was only curl -X POST.. now there is some logic. Maybe it's better (since it's py-sdk and we already have python) to implement this using standard python packages, e.g. urlib or http?

File: proto/Makefile
Comment (ig-ra) [inline proto/Makefile]: this should be run in venv, since on most distributions pip won't allow install in user dir.
I suggest to use poetry or venv we are using for python testing

File: proto/Makefile
Comment (ig-ra) [inline proto/Makefile]: should we manage everything + versions via go.mod/sum? I feel uncomfortable with `go install` here

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go]: done

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go]: refactored, irrelevant

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go:52]: I'm not sure about host:port, it can be any valid ip or dns
but I do try to connect to each of them in the configure function and send health RPC and fail if it doesn't work

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go]: done

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go:57]: I tend to not change stuff on personal preferences unless they are convention or has real world implications, 
is this one of them ?

File: cmd/ak/cmd/runtimes/runtime.go
Comment (efiShtain) [reply cmd/ak/cmd/runtimes/runtime.go:42]: I just don't understand what's the purpose of local, what does it mean ?
if #663  solves it, lets do that after we merge this

File: internal/backend/temporalclient/config.go
Comment (efiShtain) [reply internal/backend/temporalclient/config.go]: done

File: internal/backend/httpsvc/intercept.go
Comment (efiShtain) [reply internal/backend/httpsvc/intercept.go]: refactored to inject mux using fx only in server init (fx.Invoke)
good idea
thanks

File: runtimes/pythonrt/pythonrt_test.go
Comment (efiShtain) [reply runtimes/pythonrt/pythonrt_test.go]: done

File: runtimes/pythonrt/pythonrt_test.go
Comment (efiShtain) [reply runtimes/pythonrt/pythonrt_test.go]: I can't use t.log since this is a different go routine,

do you have a workaround for it ?

File: runtimes/pythonrt/pythonrt_test.go
Comment (efiShtain) [reply runtimes/pythonrt/pythonrt_test.go]: done

File: runtimes/pythonrt/runner/remote_pb2_grpc.py
Comment (efiShtain) [reply runtimes/pythonrt/runner/remote_pb2_grpc.py]: fixed

File: runtimes/pythonrt/pythonrt_test.go
Comment (efiShtain) [reply runtimes/pythonrt/pythonrt_test.go]: solved it with a channel and time.After, giving the server 2 seconds to start
should be more than enough, its just the worker PB, not full AK

let me know if that works for you or you have another option

```go
func setupServer(l *zap.Logger) (*http.Server, error) {
	mux := &http.ServeMux{}
	ConfigureWorkerGRPCHandler(l, mux)
	server := &http.Server{
		Addr:    "localhost:9980",
		Handler: h2c.NewHandler(mux, &http2.Server{}),
	}

	if err := ConfigureLocalRunnerManager(l, LocalRunnerConfig{WorkerAddress: "localhost:9980"}); err != nil {
		return nil, err
	}

	errChan := make(chan error)

	go func() {
		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
			errChan <- err
			return
		}
	}()

	select {
	case err := <-errChan:
		return nil, err
	case <-time.After(2 * time.Second): // give server time to start
		return server, nil
	}
}
```

File: runtimes/pythonrt/runner_manager_local.go
Comment (efiShtain) [reply runtimes/pythonrt/runner_manager_local.go]: done

File: runtimes/pythonrt/runner_manager_local.go
Comment (efiShtain) [reply runtimes/pythonrt/runner_manager_local.go]: done  + added check loop

File: runtimes/pythonrt/worker_grpc_handler.go
Comment (efiShtain) [reply runtimes/pythonrt/worker_grpc_handler.go:95]: why?
isn't it over engineering it ?

File: .github/workflows/ci-proto.yml
Comment (efiShtain) [reply .github/workflows/ci-proto.yml:25]: ENG-1559
will handle in the future
it's not that simple and requires code changes 

naive try changes the code generation too much (introduce connect, break linting and more)

File: runtimes/pythonrt/py-sdk/autokitteh/google.py
Comment (efiShtain) [reply runtimes/pythonrt/py-sdk/autokitteh/google.py]: so 2 or 3 ? what's the correct value ?

File: runtimes/pythonrt/py-sdk/autokitteh/google.py
Comment (efiShtain) [reply runtimes/pythonrt/py-sdk/autokitteh/google.py]: done

File: runtimes/pythonrt/testdata/pycache/__pycache__/handler.cpython-312.pyc
Comment (efiShtain) [reply runtimes/pythonrt/testdata/pycache/__pycache__/handler.cpython-312.pyc:1]: removed, the entire directory is in gitignore, but git doesn't remove stuff if it's already in git and you add it to ignore

next time the file won't be added

File: runtimes/pythonrt/testdata/trigger_webhook.go
Comment (efiShtain) [reply runtimes/pythonrt/testdata/trigger_webhook.go:1]: this is a test utility @tebeka used

I don't see any problem with it, but wdyt ?

File: .github/workflows/ci-proto.yml
Comment (efiShtain) [reply .github/workflows/ci-proto.yml]: I can switch to naive installation of protoc, but I don't see a problem with this 
if you prefer to switch, please open a ticket and assign it to me

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go:22]: ENG-1560

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go:19]: ENG-1561

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go]: removed, not well known

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go]: removed

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go:48]: I'll rethink it
my problem is basically with the local runs

this part, makes me pass stuff I'm not sure I need to pass

```go
func runtimes() sdkservices.Runtimes {
	if local {
		// TODO: what is local ? what to do about logger ? which config ?
		l := zap.New(nil)
		return kittehs.Must1(backendRuntimes.New(&backendRuntimes.Config{LazyLoadLocalVEnv: true}, l, nil))
	}
	return common.Client().Runtimes()
}

```

File: backend/runtimes/runtimes.go
Comment (efiShtain) [reply backend/runtimes/runtimes.go]: correct, fixed

File: proto/autokitteh/remote/v1/remote.proto
Comment (efiShtain) [reply proto/autokitteh/remote/v1/remote.proto:11]: future stuff
once we have more configs

File: proto/autokitteh/remote/v1/remote.proto
Comment (efiShtain) [reply proto/autokitteh/remote/v1/remote.proto:24]: ENG-1562

File: proto/autokitteh/remote/v1/remote.proto
Comment (efiShtain) [reply proto/autokitteh/remote/v1/remote.proto:1]: the diagram is part of the workflow isolation design documents
everything is described there in the mermaid diagram

do we need to add it here as well  ?

File: runtimes/pythonrt/build.go
Comment (efiShtain) [reply runtimes/pythonrt/build.go:17]: ENG-1563
refactor build system to use remote runners / builders

File: backend/runtimes/runtimes.go
Comment (tebeka) [reply backend/runtimes/runtimes.go:52]: OK

File: runtimes/pythonrt/worker_grpc_handler.go
Comment (tebeka) [reply runtimes/pythonrt/worker_grpc_handler.go:95]: There's a lot of code duplication, but main I'm concerened that someone will miss locking. If you access via API you'll have locking always.

File: internal/backend/httpsvc/intercept.go
Comment (tebeka) [reply internal/backend/httpsvc/intercept.go]: As the Mandalorian say: Why not?

File: runtimes/pythonrt/local_runner_test.go
Comment (tebeka) [reply runtimes/pythonrt/local_runner_test.go:186]: That's an old comment, we can delete it here.

**Comment by tebeka** [comment]
> I defer to @tebeka as long as this code is enabled by a feature flag, and not just replace what we have now.
That's not something we have right now, we went all or nothing. If we see issues we can revert to the old one.
If you want, we can use build tags for the new runtime, but it'll take more development effort.

File: proto/Makefile
Comment (ig-ra) [reply proto/Makefile]: plz open ticket to fix py-deps

File: backend/runtimes/runtimes.go
Comment (daabr) [reply backend/runtimes/runtimes.go:57]: It's not a personal preference, it's good style. Multi-line `if` statements are confusing and hard to understand. The shortcut of joining `x := y` and `if x` into `if x := y; x` is useful only if it fits comfortably in a single line. If not, it's better to keep them separate, especially if `y` is multi-line on its own.

File: runtimes/pythonrt/py-sdk/autokitteh/google.py
Comment (daabr) [reply runtimes/pythonrt/py-sdk/autokitteh/google.py]: 3 is correct. I have a task to add a unit test for this, but too many urgent things before that.

File: proto/Makefile
Comment (daabr) [reply proto/Makefile]: I share your discomfort and I suspect this is hiding problems, but I can't find any specific issue at this time. Either way, go.mod is managed automatically with tooling. If `go mod tidy` doesn't this it needs to add it to go.mod, adding it manually won't make things better.

File: proto/Makefile
Comment (efiShtain) [reply proto/Makefile]: there is a ticket to use buf to generate python binding for remote runtime, 
I think once it works I can remove all of this dependencies

I'll add a reference in the ticket to this discussion

File: proto/Makefile
Comment (efiShtain) [reply proto/Makefile]: same as py deps

should be solved once we move to use buf to generate the bindings,

**Review by daabr — CHANGES_REQUESTED** [review]
I committed a few small-and-obvious fixes.

I also created stacked PR #681 - critical in my opinion, but not obvious, and not finished. Let's discuss the TODOs in the PR description ASAP.

In addition, these are bugs that I observed that need to be fixed before merging this PR:

1. "Force stop sessions" leaves the Python process behind

2. Python is writing directly to stdout - this is wrong, and can mess up Datadog ingestion

3. https://github.com/autokitteh/autokitteh/pull/660/commits/66292f05d7409f94f5e56991bda6023dcf891bcb - this is wrong, we need to use the actual logger config, not any preconfigured defaults

Lastly, ENG-1559 is critical to fix urgently! If not in this PR then right after it:
   - This PR is disabling Gemini which breaks the [slack support](https://github.com/autokitteh/kittehub/blob/main/slack_support) project
   - This PR creates runtime warnings:

```
…/autokitteh/venv/lib/python3.12/site-packages/google/protobuf/runtime_version.py:112: UserWarning: Protobuf gencode version 5.27.2 is older than the runtime version 5.28.2 at remote.proto. Please avoid checked-in Protobuf gencode that can be obsolete.
```

**Comment by efiShtain** [comment]
> I committed a few small-and-obvious fixes.
> 
> I also created stacked PR #681 - critical in my opinion, but not obvious, and not finished. Let's discuss the TODOs in the PR description ASAP.
> 
> In addition, these are bugs that I observed that need to be fixed before merging this PR:
> 
> 1. "Force stop sessions" leaves the Python process behind
> 2. Python is writing directly to stdout - this is wrong, and can mess up Datadog ingestion
> 3. [66292f0](https://github.com/autokitteh/autokitteh/commit/66292f05d7409f94f5e56991bda6023dcf891bcb) - this is wrong, we need to use the actual logger config, not any preconfigured defaults
> 
> Lastly, [ENG-1559](https://linear.app/autokitteh/issue/ENG-1559) is critical to fix urgently! If not in this PR then right after it:
> 
> * This PR is disabling Gemini which breaks the [slack support](https://github.com/autokitteh/kittehub/blob/main/slack_support) project
> * This PR creates runtime warnings:
> 
> ```
> …/autokitteh/venv/lib/python3.12/site-packages/google/protobuf/runtime_version.py:112: UserWarning: Protobuf gencode version 5.27.2 is older than the runtime version 5.28.2 at remote.proto. Please avoid checked-in Protobuf gencode that can be obsolete.
> ```


1 - can you show me how to reproduce stop sessions not working?
2 - what do you suggest about python logs ? I plan to convert them to standard output
3 - agreed, once runtime config PR is merged, I'll fix it and get the correct logger

why do you think 1559 is so critical?

I'm checking the warnings

**Review by daabr — APPROVED** [review]

**Pull request merged**
